filament/backend/src/JobQueue.h

```diff
@@ -0,0 +1,271 @@
+
+    /**
+     * Pushes a new job into queue.
+     *
+     * If the queue is in the process of shutting down (via a call to `stop`), this method does
+     * nothing (a no-op) and returns an invalid job ID.
+     *
+     * @param job The function/lambda to be executed.
+     * @param preIssuedJobId The previously issued job ID where this job is assigned to.
+     *  If the value is `InvalidJobId`, a new job ID is generated internally.
+     * @return A new job ID. If a `preIssuedJobId` is provided, that specific ID is returned instead.
+     */
+    JobId push(Job job, JobId preIssuedJobId = InvalidJobId);
+
+    /**
+     * Retrieves the next job from queue.
+     *
+     * This method returns a valid job ID if there are pending jobs in the queue, even if the queue
+     * is currently in the process of shutting down (via a call to `stop`).
+     *
+     * @param shouldBlock If true (typically used by ThreadWorker), waits for a job and returns it.
+     *  If false (typically used by AmortizationWorker), tries retrieving a job. But may return an
+     *  empty job if there's no pending job.
+     * @return The next job. If an empty job is returned, it has different meaning depending on the
+     *  `shouldBlock` value:
+     *    - true: Stop processing (shutting down).
+     *    - false: No jobs currently in queue.
+     */
+    Job pop(bool shouldBlock);
+
+    /**
+     * Retrieves a batch of next jobs from queue. Always non-blocking.
+     *
+     * This method returns a valid job ID if there are pending jobs in the queue, even if the queue
+     * is currently in the process of shutting down (via a call to `stop`).
+     *
+     * @param maxJobsToPop The maximum number of jobs to retrieve.
+     *  If < 0, retrieves all pending jobs.
+     * @return A FixedCapacityVector<Job> containing the retrieved jobs.
+     * Returns an empty vector if the queue is empty.
+     */
+    utils::FixedCapacityVector<Job> popBatch(int maxJobsToPop);
+
+    /**
+     * Generate a new job ID. This newly generated ID is meant to be used for the `preIssuedJobId`
+     * parameter of `push` method.
+     *
+     * @return A new job ID.
+     */
+    JobId issueJobId() noexcept;
+
+    /**
+     * Cancels a job by its ID.
+     *
+     * @param jobId The job ID to cancel.
+     * @return true if the job was found and cancelled, false otherwise.
+     */
+    bool cancel(JobId jobId) noexcept;
+
+    /**
+     * Signals the queue to shut down, after which no further jobs can be added using the `push`
+     * method. but all jobs already pushed can still be processed using `pop`.
+     */
+    void stop() noexcept;
+
+private:
+    JobQueue(const JobQueue&) = delete;
+    JobQueue& operator=(const JobQueue&) = delete;
+
+    std::mutex mQueueMutex;
+    std::condition_variable mQueueCondition;
+    std::unordered_map<JobId, Job> mJobsMap;
+    std::queue<JobId> mJobOrder;
+    JobId mNextJobId = 0;
```

Comment:
my understanding is that `JobId` are not recycled, so we're relying on mNextJobId to never overflow/wrap around? Should we use a int64 then? 
I think at least we should have an assert when we do the++, or document in a comment what the expectation is here.